R/function-isotope.R

```diff
@@ -0,0 +1,149 @@
+#' Title Isotope identification in spectra
+#'
+#' @description
+#' Given a spectrum (i.e. a peak matrix with m/z and intensity values)
+#' the function identifies groups of peaks that correspond to isotopes
+#'
+#' @param x `matrix` with spectrum data (columns `mz` and `intensity`).
+#'
+#' @param isotopeDefinition `matrix` with isotopes definition (columns `mzd`
+#' and `maxint`). This matrix has to have rows ordered according to the column
+#' `mzd` values in order for each returned group to be ordered.
+#'
+#' @param ppm numeric(1) representing a relative, value-specific
+#' parts-per-million (PPM) tolerance for the relaxed matching of m/z values of
+#' peaks
+#'
+#' @param seed_mz numeric ordered vector containing m/z values. If provided,
+#' the function checks if there are peaks in x whose m/z match them. If so,
+#' it looks for isotope groups on this subsets of peaks.
+#'
+#' @param duplicates character(1), how to handle duplicated matches. Has to be
+#' one of c("keep", "closest", "remove"). No abbreviations allowed.
+#'
+#' @return list of vectors. Each vector in the returned list contains the
+#' indexes of the rows in `x` that match a isotope.
+#'
+#' @details The function iterates over the rows of `x`. For the `i`-th m/z
+#' value in `x` it computes m/z of would-be isotope peaks by adding the `mzd`s
+#' in `isotopeDefinition`. Possible matches of those with the other m/z values
+#' in `x` are then searched. If any is found and the ratio between the
+#' intensities of the corresponding peaks and the `i`-th peak is smaller than a
+#' threshold (specified in the `maxint` column of
+#' `isotopeDefinition`), their indexes are grouped together.
+#'
+#' It can happen that some index is present in more than one group. The user can
+#' decide what to do in this case by choosing the parameter `duplicates`.
+#' `duplicates = "keep"` leaves the groups as they are.
+#' `duplicates = "closest"` removes from the list groups for which another group
+#'  which shares some indexes with it exists and constitutes a better match.
+#' `duplicates = "remove"` removes from the list groups for which another group
+#'  which shares some indexes with it exists
+#'
+#' @examples
+#'
+.isotope_peaks <- function(x, isotopeDefinition = isotopeDefinition(), ppm = 20,
+                           seed_mz = NULL, duplicates = "keep") {
+
+  if(duplicates != "keep" && duplicates != "closest" && duplicates != "remove")
+    stop("'duplicates' has to be one of \"keep\", \"closest\" or \"remove\".")
+
+  lst <- list()
+  idxs_o <- which(x[, 2] > 0)
+  xred <- x[idxs_o, ]
+  if (is.null(seed_mz)) idxs <- 1:nrow(xred)
+  else idxs <- na.omit(MsCoreUtils::closest(seed_mz, xred[, 1], tolerance = 0,
+                                            ppm = ppm, duplicates = "closest"))
+  sgd <- NULL
+  for (i in idxs)
+  {
+    append_to_current <- FALSE
+    for (j in 1:nrow(isotopeDefinition))
```

Comment:
Do we need this loop? wouldn't it work if you did:

```r
tmp <- x[i, 1] + isotopeDefinition[, 1]
cls <- closest(tmp, x[, 1], tolerance = 0, ppm = ppm)
```

That way (I guess) you could also get around the `append_to_current`.

Reply:
Yes, I think that would work. I don't know why I did in that strange way. Thank you!